OpenRA.Mods.Common/Traits/World/EditorCursorLayer.cs

```diff
@@ -12,125 +12,41 @@
-			world = self.World;
-			editorLayer = self.Trait<EditorActorLayer>();
-			terrainRenderer = self.Trait<ITiledTerrainRenderer>();
-
-			Type = EditorCursorType.None;
-		}
+		static readonly IEnumerable<IRenderable> NoRenderables = Enumerable.Empty<IRenderable>();
 
-		void IWorldLoaded.WorldLoaded(World w, WorldRenderer wr)
+		public void SetBrush(IEditorBrush brush)
 		{
-			resourceRenderers = w.WorldActor.TraitsImplementing<IResourceRenderer>().ToArray();
+			this.brush = brush;
 		}
 
 		void ITickRender.TickRender(WorldRenderer wr, Actor self)
 		{
 			if (wr.World.Type != WorldType.Editor)
 				return;
 
-			if (Type == EditorCursorType.TerrainTemplate || Type == EditorCursorType.Resource)
-			{
-				var cell = wr.Viewport.ViewToWorld(Viewport.LastMousePos);
-				if (terrainOrResourceCell != cell || terrainOrResourceDirty)
-				{
-					terrainOrResourceCell = cell;
-					terrainOrResourceDirty = false;
-					terrainOrResourcePreview.Clear();
-
-					var pos = world.Map.CenterOfCell(cell);
-					if (Type == EditorCursorType.TerrainTemplate)
-						terrainOrResourcePreview.AddRange(terrainRenderer.RenderPreview(wr, TerrainTemplate, pos));
-					else
-						terrainOrResourcePreview.AddRange(resourceRenderers.SelectMany(r => r.RenderPreview(wr, ResourceType, pos)));
-				}
-			}
-			else if (Type == EditorCursorType.Actor)
-			{
-				// Offset mouse position by the center offset (in world pixels)
-				var worldPx = wr.Viewport.ViewToWorldPx(Viewport.LastMousePos) - wr.ScreenPxOffset(actorCenterOffset);
-				var cell = wr.Viewport.ViewToWorld(wr.Viewport.WorldToViewPx(worldPx));
-				var subCell = actorSharesCell ? editorLayer.FreeSubCellAt(cell) : SubCell.Invalid;
-				var updated = false;
-				if (actorLocation != cell)
-				{
-					actorLocation = cell;
-					Actor.ReplaceInit(new LocationInit(cell));
-					updated = true;
-				}
-
-				if (actorSubCell != subCell)
-				{
-					actorSubCell = subCell;
-
-					if (Actor.RemoveInits<SubCellInit>() > 0)
-						updated = true;
-
-					var subcell = world.Map.Tiles.Contains(cell) ? editorLayer.FreeSubCellAt(cell) : SubCell.Invalid;
-					if (subcell != SubCell.Invalid)
-					{
-						Actor.AddInit(new SubCellInit(subcell));
-						updated = true;
-					}
-				}
-
-				if (updated)
-					Actor = new EditorActorPreview(wr, null, Actor.Export(), Actor.Owner);
-			}
+			brush?.TickRender(wr, self);
 		}
 
-		static readonly IEnumerable<IRenderable> NoRenderables = Enumerable.Empty<IRenderable>();
 		IEnumerable<IRenderable> IRenderAboveShroud.RenderAboveShroud(Actor self, WorldRenderer wr)
 		{
 			if (wr.World.Type != WorldType.Editor)
 				return NoRenderables;
 
-			if (Type == EditorCursorType.TerrainTemplate || Type == EditorCursorType.Resource)
-				return terrainOrResourcePreview;
-
-			if (Type == EditorCursorType.Actor)
-				return Actor.Render().OrderBy(WorldRenderer.RenderableZPositionComparisonKey);
+			if (brush != null)
+				return brush.RenderAboveShroud(self, wr);
```

Comment:
Why not `brush?.RenderAboveShroud` to match `TickRender`?

Reply:
Mostly because this is a return statement and it isn't as clean, But sure